sw/device/tests/crypto/aes_gcm_functest.c

```diff
@@ -0,0 +1,181 @@
+ */
+static uint32_t kCiphertext256[8] = {
+    // Ciphertext =
+    // 4e6d3a963b076ba0945d29aa836f29b0fa06cdd575aab8233f1df93e80163371
+    0x963a6d4e, 0xa06b073b, 0xaa295d94, 0xb0296f83,
+    0xd5cd06fa, 0x23b8aa75, 0x3ef91d3f, 0x71331680};
+
+static const aes_gcm_test_t aes_gcm_testvectors[3] = {
+    // Empty input, empty aad, 12-bit IV, 128-bit key
+    {
+        .key_len = kAesKeyLen128,
+        .iv_len = 12,
+        .iv =
+            {// IV = 22294cae82d82e44427dfcc3
+             0xae4c2922, 0x442ed882, 0xc3fc7d42, 0x0},
+        .plaintext_len = 0,
+        .plaintext = NULL,
+        .aad_len = 0,
+        .aad = NULL,
+        .expected_tag =
+            {
+                // Tag = b7aa223a6c75a0976633ce79d9fddf06
+                0x3a22aab7,
+                0x97a0756c,
+                0x79ce3366,
+                0x06dffdd9,
+            },
+        .expected_ciphertext = NULL,
+    },
+
+    // Empty input, empty aad, 16-bit IV, 128-bit key
+    {
+        .key_len = kAesKeyLen128,
+        .iv_len = 16,
+        .iv =
+            {// IV = 22294cae82d82e44427dfcc33bacdbec
+             0xae4c2922, 0x442ed882, 0xc3fc7d42, 0xecdbac3b},
+        .plaintext_len = 0,
+        .plaintext = NULL,
+        .aad_len = 0,
+        .aad = NULL,
+        .expected_tag =
+            {// Tag = 4c59f0d420d9eb8669c40ad23b5419ba
+             0xd4f0594c, 0x86ebd920, 0xd20ac469, 0xba19543b},
+        .expected_ciphertext = NULL,
+    },
+
+    // 16-bit IV, 256-bit key, real message and aad
+    {
+        .key_len = kAesKeyLen256,
+        .iv_len = 16,
+        .iv =
+            {// IV = c58aded2e1bbecba8b16a5757e5475bd
+             0xd2de8ac5, 0xbaecbbe1, 0x75a5168b, 0xbd75547e},
+        .plaintext_len = kPlaintextLen,
+        .plaintext = kPlaintext,
+        .aad_len = kAadLen,
+        .aad = kAad,
+        .expected_tag =
+            {// Tag = 324895b3d2f656e4fa2f8ce056137061
+             0xb3954832, 0xe456f6d2, 0xe08c2ffa, 0x61701356},
+        .expected_ciphertext = kCiphertext256,
+    },
+};
+
+OTTF_DEFINE_TEST_CONFIG();
+bool test_main(void) {
+  for (size_t i = 0; i < ARRAYSIZE(aes_gcm_testvectors); i++) {
+    LOG_INFO("Starting AES-GCM test %d of %d...", i + 1,
+             ARRAYSIZE(aes_gcm_testvectors));
+    aes_gcm_test_t test = aes_gcm_testvectors[i];
+
+    // Construct key shares by setting first share to full key and second share
+    // to 0. (Note: this is not a secure construction! But it makes debugging
+    // tests easier because there is only one thing to print.)
+    uint32_t *key_shares[2];
+    uint32_t share1[8] = {0};
+    key_shares[1] = share1;
+    if (test.key_len == kAesKeyLen128) {
+      key_shares[0] = kKey128;
+    } else if (test.key_len == kAesKeyLen256) {
+      key_shares[0] = kKey256;
+    } else {
+      LOG_ERROR("No key available for key length.");
+      return false;
+    }
+
+    // Call AES-GCM encrypt.
+    uint8_t actual_ciphertext[test.plaintext_len];
```

Comment:
FYI, I've heard, but not verified, that variable-length arrays are an optional feature for C11 compilers. CI passed, so our toolchain seems to support it :)